lib/events.js

```diff
@@ -647,3 +652,84 @@ function once(emitter, name) {
     emitter.once(name, eventListener);
   });
 }
+
+const AsyncIteratorPrototype = ObjectGetPrototypeOf(
+  ObjectGetPrototypeOf(async function* () {}).prototype);
+
+function createIterResult(value, done) {
+  return { value, done };
+}
+
+function on(emitter, event) {
+  const unconsumedEvents = [];
+  const unconsumedPromises = [];
+  let error = null;
+
+  const iterator = ObjectSetPrototypeOf({
+    next() {
+      if (error) {
+        return PromiseReject(error);
+      }
+
+      const value = unconsumedEvents.shift();
+      if (value) {
+        return PromiseResolve(createIterResult(value, false));
+      }
+
+      return new Promise(function(resolve, reject) {
+        unconsumedPromises.push({ resolve, reject });
+      });
+    },
+
+    return() {
+      emitter.removeListener(event, eventHandler);
```

Comment:
it's possible for there to be pending `unconsumedEvents` in the iterator. consider `unconsumedEvents.length = 0` or `unconsumedEvents = null`